### ISSUE TITLE
Question about suitability of PyMunk for Gym

### ISSUE THREAD

**Issue by jkterry1** [viblo/pymunk#198]
Hey, I'm the maintainer of [Gym](https://github.com/openai/gym), the largest library for reinforcement learning. Right now, we depend on box2d, which is no longer maintained, because the maintainer sort of gave up and no one has volunteered to replace them in the following 5 years. Because of this, we're currently discussing rewriting these environments with PyMunk for physics, but before doing this I just wanted to check and make sure that you intend to maintain PyMunk for a good while.

I also have another question that I couldn't find anywhere- what architectures do you include wheels for/support compilation for?

**Comment by viblo** [comment]
Not sure if you saw it, but on conda there are osx-arm64 wheels. 

Anyway, the situation on Github Actions is that ARM64 is supported as a target, but the actual machines used for running are not (instead you can self host the runner). 

For MacOS it seems like I can just enable arm64 in cibuildwheel and the code will be cross compiles, but it wont be able to run the tests. On the other hand this should be offset a bit by the conda builds which do run the tests.. 

For linux arm64 it seems like the main issue is huge increase in build times since it has to be run with emulation, but seems possible.

**Comment by viblo** [comment]
I will do some experiments.
I think at least MacOS arm64 is quite nice to have precompiled. I imagine the linux users are more used to compiling things themselves, so not as important there, but if it just works then I will try. 

Windows arm is not build by cffi, until that happens I dont think its any use to try and do Pymunk.

**Comment by viblo** [comment]
Ok, I have tried with arm64 for macos. Unfortunately it did not work out of the box. 

When I looked closer it seems to be something with cffi. I noticed that there are actually not any prebuilt wheels for cffi in pypi for arm64 for macos, only linux, so maybe some issue when building it (but I know that both cffi and pymunk works on arm64/macos when its installed locally on a real mac). Anyway, I will raise an issue in cibuildwheel for this and see if that leads anywhere.

While reading around I also saw an issue in the cffi-repo about windows arm. So right now it seems like its not supported, but someone was eager to make it work so I guess its just a matter of time :)

**Comment by viblo** [comment]
Reported here https://github.com/pypa/cibuildwheel/issues/813

**Comment by jkterry1** [comment]
Has there been any new with regards to arm builds?

**Comment by viblo** [comment]
Just made a new release of Pymunk, 6.2.1, with wheels for osx arm64.

**Comment by jkterry1** [comment]
Thanks a ton!

**Comment by jkterry1** [comment]
@viblo I know it's half a decade later, but I'm trying to figure out what 2D physics engine to use within the Farama ecosystem (it's grown quite a lot - https://farama.org/) and was curious as to your take. 

Originally I was planning to port everything we still run on Box2D over to Pymunk due to issues with our Box2D bindings going from mostly to fully unmaintained in the past year (see https://github.com/Farama-Foundation/Gymnasium/issues/1597 and if you look at the pyproject.toml we're having to build stuff from source for Python 3.14) and was just coming here to tell you, but then I saw your comments on the underlying Chipmunk C repo being semi unmaintained, looked at that repo, and it both is unmaintained and if anything its degree of maintenance has decreased since this post to the point of there being almost none (and its documentation site is seemingly even down), which obviously doesn't make it the ideal choice.

On the other hand, the underlying Box2D C physics library is still fairly actively maintained to date, however there are no good Python bindings. The only ones with updates in the past year I can find are these two repos, both with only 4 stars and no long term history of maintenance:

https://github.com/DerThorsten/pyb2d3
https://github.com/giorgosg/box2d-py

-Do you have thoughts on what you would do if you were in our position of maintaining Python packages that are installed millions to hundreds of millions of times that need to depend on a long term 2D physics engine and that are, for instance, extremely widely used in university courses for students?

-If you aren't aware of any other alternative to this, and I frankly hope you can suggest a better plan than this, is switching your long term work of maintaining Python bindings for Chipmunk to maintaining Python bindings for Box2D something you might be interested in, given that you appear to have the interest in maintaining 2D physics engine Python bindings very long term but the physics engine underpinning this project has not actively been maintained in ~7 years while Box2D continues to get fairly frequent updates even now? 

If that actually is an endeavor you'd be interested in, you very much wouldn't be starting from scratch given the large number of existing unmaintained Python bindings for Box2D. If you wanted to, you also could eventually suggest that users of PyMunk to migrate over to your Box2D bindings given that PyMunk has a relatively widespread user base, especially within the PyGame community, but through some flavor of misfortune and misblowing of the tea leaves presently doesn't have a well maintained physics engine underpinning it.

**Comment by jkterry1** [comment]
I edited that post in case you just read the email notification

**Comment by jkterry1** [comment]
Looking at your repo further, you appear to have gone the other way with this problem and effectively taken over maintaining Chipmunk yourself in the past year at https://github.com/viblo/Munk2D , I believe meaning that we should be good to just proceed with migrating all of our tools over to Pymunk as I was originally planing?

**Comment by jkterry1** [comment]
Then I suppose one final question - assuming we are going to fully migrate all of our tooling over to Pymunk as I suspect given my last post, would you be willing to transfer ownership of the Github repos for Pymunk and Munk2D to an open source non profit of your choosing so that the repos remain maintainable should you get hit by a bus or similar? You could transfer them to us if you wanted (2D physics engines aren't exactly our thing but our dependency on one is such that I'm cool with it if you'd like to), but I could also introduce you to the Linux Foundation, Mozilla, or another independent name brand open source nonprofit of your choosing should you have a preferred one.

**Comment by jkterry1** [comment]
Additionally, while I've had no direct interactions with any members, from what I've publicly read about the new PyGame-Community organization, how it's come about, and how it's been run, I at least suspect that they could possibly be a well run and natural fit to keeping this repo from having a bus factor of 1 and potentially might be enthusiastic to have Pymunk officially join them given how much their community uses it.

**Comment by viblo** [comment]
Nice to see you again!

Right, as you saw I felt the best way forward for Pymunk was to make a fork of Chipmunk. I cant say the fork has received much in maintenance yet, but a few things it does have, at least around infrastructure like how to build it. I have also done some experiments in a separate fork (https://github.com/viblo-aidev/Munk2d) where I want to try some performance improvements. 

As for Box2D, I have only considered it very briefly. I do think the new v3 is very interesting technically, and its now in C just like Chipmunk. The problem is I think it would change quite a lot in the API and how the simulations behave..

**Comment by viblo** [comment]
Regarding the bus factor, I do think its a valid concern. But I don't think moving the ownership will help or is something I am interested in at this point.
  
I am of course aware of the problems that do happen from time to time. To take an example, just the other day I saw that the popular `httpx` library has become unmaintained. That was "solved" by Pydantic forking it to a renamed `httpx2` which I think is a reasonable outcome. E.g. just having an Open Source license is enough to guard against such cases, as long as there people around willing to take over. 

However, the bigger issues I have seen is when either the situation is unclear, there's no-one to take over or the most troublesome, when there's serious disagreement between the maintainers. In my impression the ownership structure doesn't really help much in those cases. I.e the several python box2d libraries, or the mess around MkDocs or how Pygame/PygameCE happened. 

In the end I think that with an open source license you get very far in terms of what you can do, but it does of course depends on that there's someone both capable and willing to do the work. And for the majority of projects its exactly this that is the main problem unfortunately.

**Comment by jkterry1** [comment]
Hey there,

I'm happy to use PyMunk with your maintained fork of Chipunk, and I'll close this issue. I greatly appreciate the long term work that you've been doing with this :)

It's your project, and I absolutely don't want to try to make you join an open source non profit if you don't want to.

For what its worth, the reason I brought up open source organizations is that there's actually kind of a lot of examples of where a project being part of an open source nonprofit has prevented a project from being abandoned or similar, despite what the license gives you for your example.

A project being part of an open source organization frequently reduces abandonment in a few ways:

1. The barrier to entry for the maintainers is much lower, because they don't have to get the community traction for their new project, or make it "new new official project", they can just keep pushing changes to the same project as before. 
2. Similarly, this also allows for projects to go through a series of maintainers in a way that forking wouldn't, each making their own contributions during their tenure.
3. Taking over an existing widely used project creates much more direct professional benefits for a maintainer are a lot more immediate,
4. With many (but not all) host organization, it gives a sense of community ownership over a project and drives people to make contributions as applicable/where needed, in turn increasing the amount of contributions and decreasing abandonment

These factors also work together synergistically. To pick an example that I personally know especially well, [PettingZoo](https://github.com/Farama-Foundation/PettingZoo), was half of my PhD thesis and is the most used open source library in a subfield of a subfield of AI. The reason it's continued so long has basically been someone in a leadership role saying something to a credibly prospective contributor the effect of "Hey man, want to maintain the most used open source library in this field of AI?" People in the position to do so often take that request, because it's an extremely professionally useful offer in many situations and/or they intrinsically want to, do so for as long as they can, and after some number of years someone new says years to the same question. The barrier of making a whole new fork and official version wouldn't let that happen. PettingZoo has now gone through 5 maintainers in the past 7 years, and that definitely wouldn't be possible in another model. Moreover throughout it's historically, particularly in the last year and around 4 years ago, it's also gotten an enormous amount of contributions from community that emerged around Farama, from people who's interest would otherwise be largely disjoint from PettingZoo.

PyMunk is more than widely used and well known enough for all those factors to work in the favor of you and the project, provided that you open source organization that has a good community, that you have a good working relationship with, and that you generally think is managed in a long term and sustainable manner. Those last factors are why I suggested that the PyGame-CE community as an option that potentially could, given a series of discussions and agreements between you and them, be a much better option than say Apache or the linux Foundation. At minimum, smaller organizations tend to outperform the large ones for projects smaller than say Kubernetes.

Another middle ground possible worth contemplating is moving the repos to a GitHub organization, creating a discord server and/or discuss forum (you can almost certainly get someone to moderate them for you if you directly ask for someone to volunteer for the role), and eventually looking to add a second owner to the GitHub organization and PyPI account. That would convey almost all of the same benefits for none of the risk of loss of control of something you've worked on for 15 years if that's the primary concern.

Regardless, its your project and I'll keep moving all of Farama's projects over to it, thats just my personal perspective on open source.
